src/unix_term.rs

```diff
@@ -66,41 +65,71 @@ pub fn terminal_size(out: &Term) -> Option<(u16, u16)> {
     }
 }
 
-pub fn read_secure() -> io::Result<String> {
-    let mut f_tty;
-    let fd = unsafe {
-        if libc::isatty(libc::STDIN_FILENO) == 1 {
-            f_tty = None;
-            libc::STDIN_FILENO
-        } else {
-            let f = fs::OpenOptions::new()
-                .read(true)
-                .write(true)
-                .open("/dev/tty")?;
-            let fd = f.as_raw_fd();
-            f_tty = Some(BufReader::new(f));
-            fd
+enum Input<T> {
+    Stdin(io::Stdin),
+    File(T),
+}
+
+impl<T> Input<T>
+where
+    T: BufRead,
+{
+    fn read_line(&mut self, buf: &mut String) -> io::Result<usize> {
+        match self {
+            Self::Stdin(s) => s.read_line(buf),
+            Self::File(f) => f.read_line(buf),
         }
+    }
+}
+
+impl AsRawFd for Input<fs::File> {
+    fn as_raw_fd(&self) -> RawFd {
+        match self {
+            Self::Stdin(s) => s.as_raw_fd(),
+            Self::File(f) => f.as_raw_fd(),
+        }
+    }
+}
+
+impl AsRawFd for Input<BufReader<fs::File>> {
+    fn as_raw_fd(&self) -> RawFd {
+        match self {
+            Self::Stdin(s) => s.as_raw_fd(),
+            Self::File(f) => f.get_ref().as_raw_fd(),
+        }
+    }
+}
+
+fn input() -> io::Result<Input<fs::File>> {
+    let stdin = io::stdin();
+    if is_a_terminal(&stdin) {
+        Ok(Input::Stdin(stdin))
+    } else {
+        let f = fs::OpenOptions::new()
+            .read(true)
+            .write(true)
+            .open("/dev/tty")?;
+        Ok(Input::File(f))
+    }
+}
```

Comment:
Should make this an `Input` constructor. Suggest using `match` instead of `if` and inlining the `f` binding. Pick good names to distinguish `input()` and `read_secure()`?

Reply:
Done.